### DISCUSSION TITLE
Uniform injection instead of random injection

### DISCUSSION THREAD

**Discussion by cooluser26** [epochpic/epoch#636] [Q&A] [answered]
Hi,

Is it possible to modify the code (easily) somehow to inject the particles uniformly, at least in the 1D case? I am implementing an exponentially decaying density profile and this causes the E-B field to explode over just ~ e-folding scale (10^6 times larger so definitely some artificial instability going on here). I think that maybe injecting the particles uniformly would be the fix? I saw that someone in the past was able to modify an older of EPOCH to get the following a smoother profile (see image). Is that still possible somehow? 

Thank you,

William.

![image](https://github.com/Warwick-Plasma/epoch/assets/108578700/35a70e27-d26e-4e05-b45b-ccef77b880a5)

**Comment by Status-Mirror** [comment, marked as answer]
Hey William,

It is possible to do this, but "easily" depends on how comfortable you are tweaking things under the hood.

The EPOCH domain is composed of simulation cells surrounded by ghost cells. Macro-particles have a "shape", such that the EM fields on the particle are interpolated using multiple nearby cells. When this shape extends beyond the simulation cells, we can use the fields in the ghost cells, but when the particle-centre drifts into a ghost-cell (and we have outflow boundary conditions), the particle is deleted.

The injectors in EPOCH work by populating these ghost cells with macro-particles describing a plasma with the parameters you specify in the injector block. These are pushed using their loaded momenta, and those which enter the simulation window during the time-step are kept, and the ones which stay in the ghost cell are deleted.

The positions of these loaded injector particles are calculated in the `run_injectors` subroutine in `src/physics_packages/injectors.F90`. The specific lines are:

```
    DO ipart = 1, parts_this_time
      CALL create_particle(new)

      new%part_pos = bdy_pos - random() * v_inject_dt
```

- `bdy_pos`: $x$ position of the simulation boundary (either `x_min` or `x_max`)
- `v_inject_dt`: Speed $v_\text{drift}$ corresponding to injector-block `drift_x`, multiplied by simulation timestep `dt`.
- `random()`: A uniformly distributed random number between 0 and 1.

These lines position macro-particles such that if they all move with $v_\text{drift}$, then they will all enter the simulation window at random times over the time-step. If you want uniform injection, these are the lines to target. My naive approach would be to replace `random()` with a function which uniformly goes from 0 to 1:

```
    DO ipart = 1, parts_this_time
      CALL create_particle(new)

      new%part_pos = bdy_pos - REAL(ipart-1, num) / REAL(parts_this_time, num) * v_inject_dt
```

Note that I do not allow this fraction to reach 1, such that we don't have the last particle on step $n$ overlap with the first particle in step $n+1$. Also, this will only provide uniform injection if all particles move at the same speed $v_\text{drift}$. You won't be able to keep a uniform separation if you set up a plasma with thermal fluctuations in momentum.

Alternatively, you could specify the injection time and momentum of all particles by using [file injectors](https://epochpic.github.io/quickstart/basic_examples/injectors.html#file-injectors). This may give you more control to play around with different distributions of particles, at the cost of writing long particle injection lists.

I hope this mini-lecture on the EPOCH injectors was useful!
Stuart

**Reply by cooluser26** [reply]
Thanks so much for the detailed answer, Stuart! Been learning a lot about EPOCH and PIC in general over the past few days from your answers! I also find that most of my "stability" revolves around boundary stuff and it seems to be fixed by the idea that "don't intialize anything too near to the boundary". I managed to go around my stability issue by initializing my "beam" a bit further away from the boundary
